packages/attachments/src/model.ts

```diff
@@ -0,0 +1,456 @@
+   */
+  readonly contentFactory: IAttachmentsModel.IContentFactory;
+
+  /**
+   * Test whether the model is disposed.
+   */
+  get isDisposed(): boolean {
+    return this._isDisposed;
+  }
+
+  /**
+   * Dispose of the resources used by the model.
+   */
+  dispose(): void {
+    if (this.isDisposed) {
+      return;
+    }
+    this._isDisposed = true;
+    this.map.dispose();
+    Signal.clearData(this);
+  }
+
+  /**
+   * Whether the specified key is set.
+   */
+  has(key: string): boolean {
+    return this.map.has(key);
+  }
+
+  /**
+   * Get an item at the specified index.
+   */
+  get(key: string): IAttachmentModel {
+    return this.map.get(key);
+  }
+
+  /**
+   * Set the value at the specified index.
+   */
+  set(key: string, value: nbformat.IMimeBundle): void {
+    // Normalize stream data.
+    let item = this._createItem({ value });
+    this.map.set(key, item);
+  }
+
+  /**
+   * Clear all of the attachments.
+   */
+  clear(): void {
+    each(this.map.values(), (item: IAttachmentModel) => { item.dispose(); });
+    this.map.clear();
+  }
+
+  /**
+   * Deserialize the model from JSON.
+   *
+   * #### Notes
+   * This will clear any existing data.
+   */
+  fromJSON(values: nbformat.IAttachments) {
+    this.clear();
+    each(Object.keys(values), (key) => { this._add(key, values[key]); });
+  }
+
+  /**
+   * Serialize the model to JSON.
+   */
+  toJSON(): nbformat.IAttachments {
+    let ret: nbformat.IAttachments = {};
+    for (let key of this.map.keys()) {
+      ret[key] = this.map.get(key).toJSON();
+    }
+    return ret;
+  }
+
+  /**
+   * Add an item to the list.
+   */
+  private _add(key: string, value: nbformat.IMimeBundle): void {
+    // Create the new item.
+    let item = this._createItem({ value });
+
+    // Add the item to our list and return the new length.
+    this.map.set(key, item);
+  }
+
+  protected map: IObservableMap<IAttachmentModel> = null;
+
+  /**
+   * Create an attachment item and hook up its signals.
+   */
+  private _createItem(options: IAttachmentModel.IOptions): IAttachmentModel {
+    let factory = this.contentFactory;
+    let item = factory.createAttachmentModel(options);
+    item.changed.connect(this._onGenericChange, this);
+    return item;
+  }
+
+  /**
+   * Handle a change to the list.
+   */
+  private _onMapChanged(sender: IObservableMap<IAttachmentModel>, args: IObservableMap.IChangedArgs<IAttachmentModel>) {
+    if (this._serialized && !this._changeGuard) {
+      this._changeGuard = true;
+      this._serialized.set(this.toJSON());
+      this._changeGuard = false;
+    }
+    this._changed.emit(args);
+    this._stateChanged.emit(void 0);
+  }
+
+  /**
+   * If the serialized version of the outputs have changed due to a remote
+   * action, then update the model accordingly.
+   */
+  private _onSerializedChanged(sender: IObservableValue, args: ObservableValue.IChangedArgs) {
+    if (!this._changeGuard) {
```

Comment:
I think this is needed to prevent an infinite recursion:
- serialized value changes
- `_onSerializedChanged`
- `fromJSON`
- `clear` and / or `_add`
- `_onMapChanged`
- `_serialized.set()` (currently prevented by guard)
- ... repeat to `_onSerializedChanged`

Reply:
Disregard this ^